internal/ihop/client.go

```diff
@@ -0,0 +1,488 @@
+	DiffID string
+
+	v1.Layer
+}
+
+type Client struct {
+	dir      string
+	docker   *docker.Client
+	keychain authn.Keychain
+}
+
+func NewClient(dir string) (Client, error) {
+	client, err := docker.NewClientWithOpts(docker.FromEnv)
+	if err != nil {
+		return Client{}, err
+	}
+
+	return Client{
+		dir:      dir,
+		docker:   client,
+		keychain: authn.DefaultKeychain,
+	}, nil
+}
+
+func (c Client) Build(def DefinitionImage, platform string) (Image, error) {
+	sum := sha256.Sum256([]byte(def.Dockerfile))
+	sess, err := session.NewSession(context.Background(), def.Dockerfile, hex.EncodeToString(sum[:]))
+	if err != nil {
+		return Image{}, err
+	}
+
+	dockerAuthProvider := authprovider.NewDockerAuthProvider(os.Stderr)
+	sess.Allow(dockerAuthProvider)
+
+	if len(def.Secrets) > 0 {
+		secretsDir, err := ioutil.TempDir("", "docker-secrets")
+		if err != nil {
+			return Image{}, err
+		}
+		defer os.RemoveAll(secretsDir)
+
+		fs := make([]secretsprovider.Source, 0, len(def.Secrets))
+		for id, secret := range def.Secrets {
+			path := filepath.Join(secretsDir, id)
+			err = os.WriteFile(path, []byte(secret), 0600)
+			if err != nil {
+				return Image{}, err
+			}
+
+			fs = append(fs, secretsprovider.Source{
+				ID:       id,
+				FilePath: path,
+			})
+		}
+
+		store, err := secretsprovider.NewStore(fs)
+		if err != nil {
+			return Image{}, err
+		}
+
+		sess.Allow(secretsprovider.NewSecretProvider(store))
+	}
+
+	go func() {
+		_ = sess.Run(context.Background(), func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) {
+			return c.docker.DialHijack(ctx, "/session", proto, meta)
+		})
+	}()
+	defer sess.Close()
+
+	contextDir := filepath.Dir(def.Dockerfile)
+	excludes, err := build.ReadDockerignore(contextDir)
+	if err != nil {
+		return Image{}, err
+	}
+
+	contextDir, relDockerfile, err := build.GetContextFromLocalDir(contextDir, def.Dockerfile)
+	if err != nil {
+		return Image{}, err
+	}
+
+	buildContext, err := archive.TarWithOptions(contextDir, &archive.TarOptions{
+		ExcludePatterns: excludes,
+		ChownOpts:       &idtools.Identity{UID: 0, GID: 0},
+	})
+	if err != nil {
+		return Image{}, err
+	}
+	defer buildContext.Close()
+
+	tag, err := randomName()
+	if err != nil {
+		return Image{}, err
+	}
+
+	tag = fmt.Sprintf("paketo.io/stack/%s", tag)
+
+	resp, err := c.docker.ImageBuild(context.Background(), buildContext, types.ImageBuildOptions{
+		BuildArgs:  opts.ConvertKVStringsToMapWithNil(def.Arguments()),
+		Dockerfile: relDockerfile,
+		NoCache:    true,
+		Remove:     true,
+		Tags:       []string{tag},
+		Version:    types.BuilderBuildKit,
+		SessionID:  sess.ID(),
+		Platform:   platform,
+	})
+	if err != nil {
+		return Image{}, nil
+	}
+	defer resp.Body.Close()
+
+	buffer := bytes.NewBuffer(nil)
+	displayChan := make(chan *client.SolveStatus)
+	go func() {
+		_ = progressui.DisplaySolveStatus(context.Background(), "", nil, buffer, displayChan)
+	}()
+
+	stream := json.NewDecoder(resp.Body)
```

Comment:
from here down, I have no idea what's going on in this function. Can you explain briefly? Is most of this taken from the old implementation?

Reply:
This code is entirely about parsing the log output coming from buildkit. If the build errors, we need to show the user the logs. These logs come in a weird JSON encoded format. You can see this output if you make a modification to one of the Dockerfiles that caused the build to fail.

Reply:
oh got it. The comments help too!